common/src/chain/transaction/signature/inputsig.rs

```diff
@@ -171,17 +171,93 @@ impl Encode for StandardInputSignature {
     }
 }
 
-// TODO: write tests
-
 #[cfg(test)]
 mod test {
     use crate::{
+        address::pubkeyhash::PublicKeyHash,
         chain::{TransactionCreationError, TxInput, TxOutput},
         primitives::{Amount, Id},
     };
 
     use super::*;
     use crypto::key::{KeyKind, PrivateKey};
+    use itertools::Itertools;
+
+    const INPUT_NUM: usize = 0;
+
+    #[test]
+    fn produce_signature_address_missmatch() {
+        let (private_key, _) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let (_, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_destination = Destination::Address(PublicKeyHash::from(&public_key));
+        let tx = generate_unsigned_tx(outpoint_destination.clone()).unwrap();
+
+        for sighash_type in sig_hash_types() {
+            println!("{sighash_type:X?}");
+
+            assert_eq!(
+                Err(TransactionSigError::PublicKeyToAddressMismatch),
+                StandardInputSignature::produce_signature_for_input(
+                    &private_key,
+                    sighash_type,
+                    outpoint_destination.clone(),
+                    &tx,
+                    INPUT_NUM,
+                )
+            );
+        }
+    }
+
+    #[test]
+    fn produce_signature_key_missmatch() {
+        let (private_key, _) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let (_, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_destination = Destination::PublicKey(public_key);
+        let tx = generate_unsigned_tx(outpoint_destination.clone()).unwrap();
+
+        for sighash_type in sig_hash_types() {
+            println!("{sighash_type:X?}");
+
+            assert_eq!(
+                Err(TransactionSigError::SpendeePrivatePublicKeyMismatch),
+                StandardInputSignature::produce_signature_for_input(
+                    &private_key,
+                    sighash_type,
+                    outpoint_destination.clone(),
+                    &tx,
+                    INPUT_NUM,
+                )
+            );
+        }
+    }
+
+    #[test]
+    fn produce_and_verify() {
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoints = [
+            Destination::Address(PublicKeyHash::from(&public_key)),
+            Destination::PublicKey(public_key),
+        ];
+
+        for (sighash_type, outpoint_destination) in
+            sig_hash_types().cartesian_product(outpoints.into_iter())
+        {
+            println!("{sighash_type:X?} {outpoint_destination:?}");
```

Comment:
I assume these prints are for debugging purposes. Two points here:
- Perhaps using a debugger is a much better way to solve such a problem. If you're failing at debugging and seeing variables being optimized out, it's because tests are by default optimized. Comment out the optimization line in the root's Cargo.toml
- I understand some times we have to do this anyway and print. If that's the case, please write a TODO comment to remind you to remove the line along the line.

Reply:
I agree that prints in tests are somewhat messy, but they aren't for debugging. In case of failure it wouldn't be obvious on what iteration it happened. With these prints it would be clear. By default (without the `--nocapture` option) `cargo test` hides output of successful tests, so there will be no mess in the output unless the tests actually failed.

The other way of handling this is to do something like that:
```rust
witness
    .verify_signature(&outpoint_destination, &sighash)
    .expect(&format!({sighash_type:X?} {outpoint_destination:?}));
```